### ISSUE TITLE
TEDIT selection by left-button sweeping?

### ISSUE THREAD

**Issue by rmkaplan** [Interlisp/medley#1517]
In Word and some (most?) other editors, a single click of the left button gives you a point selection, the caret blinking between 2 characters.  If you want to extend the selection, you hold the mouse down after clicking and sweep out the characters you want to include in the selection.

If you left-click in Tedit, you get a blinking caret, but you also see that the single character either before or after the mouse position is selected and underlined.  If you don't release the mouse, but drag it along, the caret moves with it, and you end with a single character selection somewhere else on the screen. You don't see that one end of the selection is pinned where the mouse first went down and then is extended either before or after that, as you drag and sweep.

The single character selection is kind of odd (except when you are pointing at an image object).  When we had keyboards with formatting keys (Bold, Italic,etc), pressing one of those keys would not only set the mode for future typing at that position, but also change the looks of that single character, whether you intended that or not.  (We see this behavior now with the softkeys that you get by loading lispusers/EDITKEYS, and you would also see this if we set up e.g. meta-B, meta-I etc as Tedit action keys.)

I think it makes more sense, first, to change the current behavior so that clicking with the left-button always (except for image objects) gives you a point selection, with no underlining.  You type and delete from there.  Any new looks that you apply would affect only new characters that you type at that point (in Tedit, those are the "caret looks").

But also, there is code in Tedit for extending selections by dragging the mouse, but currently this is tied only to the right button.  If you have selected a point/character with Left or a word with Middle (or a line or paragraph from the margin), then clicking and holding the right button will extend that selection  either forwards or backwards (to the character, word, paragraph boundaries according to the original type).

I don't think it would be too difficult (famous last words) to bind that extension behavior to the Left and Middle buttons, so that moving the mouse while Left or MIddle is held down doesn't change the initial point of a selection, as it does now, but extends the selection forward or backward from that anchor position.

This would make Tedit behave a little less idiosyncratically.  Of more importance, it would probably make it less dependent on having or simulating a three-button mouse.  (Interpreting double and triple clicks for determining the selection-type would be the next move in that direction.)

**Comment by nbriggs** [comment]
The first thing that comes to mind is how such a change would interact with shift (copy) select -- if you shift-click do you get nothing, because it's selecting a point, instead of the single character?

In trying various shift(copy)-select operations and also control-shift(move)-select operations, I get unexpected behavior when the target is not within the source TEdit window.

**Comment by rmkaplan** [comment]
In 4th round, where I am playing at least with the point-selection part of this, copy-looks (meta-select) still selects and underlines a single item, as before. If sweeping is implemented, I’m not sure whether a single (unextended) click should also pick out an item or do nothing unless there is a small little extension sweep.  (I think other editors without the copy-select idea use the sweep/clipboard copy/paste metaphor for copy and moving.)

Now that you make the observation, the release system also has quirky behavior when the target is not in the Tedit window.  With the shift/meta (move) down, I see the black highlighting in the Tedit window, and the highlighting stays there when the mouse comes up.  But nothing appears in the target.  However, if I then go back and click in the Tedit window, the highlighted stuff disappears.  Apparently, Tedit delays but finally executes the deletion part of the move, but nothing showed up on the target side.  Is that what you see?

So I think foreign-target behavior is a separate glitch, probably there forever.  When going to a foreign target, I suppose that Tedit should decompose its move operation into the normal copy followed by a normal delete. (Not sure why it doesn’t do that already.)

**Comment by nbriggs** [comment]
> Apparently, Tedit delays but finally executes the deletion part of the move, but nothing showed up on the target side.  Is that what you see?

Yes.  I haven't tried any of the meta- operations, only shift- and ctrl-.
When the source and target are *both* TEdit it works mostly.  It gets confused, though, if you have shift-ctrl moved the last piece of text in a window (I was selecting a whole line) and then go back to that window -- the caret is blinking in a position that looks as though it is beyond the end of the file, and attempting to shift-ctrl select back from the first target to the first source produced an "ARG NOT PIECE / LASTPIECE" from TEDIT.DO.BLUEPENDINGDELETE

**Comment by rmkaplan** [comment]
I pushed what I hope will "do the right thing" when ctrl-shifting (move) to a foreign target.  The text should appear in the non-Tedit target, and the text should disappear from the Tedit window. LMK

**Comment by nbriggs** [comment]
Yes, the ctrl-shift (move) to a foreign target works well now.

**Comment by rmkaplan** [comment]
I haven’t been able to replicate this behavior.  Is it still happening?

**Comment by nbriggs** [comment]
I haven't been able to replicate it either -- but I'll keep trying.  I guess if it fails the right thing to save for debugging is the VM.

**Comment by nbriggs** [comment]
With the most up-to-date 4th round TEdit, how do you shift-select a single character?

As I was playing around with the middle and left button I got an unexpected end-of-file.
I'm attaching the lisp.virtualmem
[lisp.virtualmem.gz](https://github.com/Interlisp/medley/files/14070102/lisp.virtualmem.gz)

**Comment by rmkaplan** [comment]
You have to extend slightly with the right button, not yet possible to just extend with the left button.
